analysis/portanalyzer.go

```diff
@@ -0,0 +1,165 @@
+package analysis
+
+import (
+	"log"
+	"math"
+
+	"github.com/sarchlab/akita/v3/sim"
+)
+
+// PortBandwidthLogger is a hook for logging messages as they go across a Port
+type PortBandwidthLogger struct {
+	sim.LogHookBase
+	sim.TimeTeller
+
+	name      string
+	startTime sim.VTimeInSec
+	endTime   sim.VTimeInSec
+	interval  sim.VTimeInSec
+
+	messageCounter      map[sim.VTimeInSec]int64
+	trafficBytesCounter map[sim.VTimeInSec]int64
+
+	logger *log.Logger
+}
+
+// NewPortMsgLogger returns a new PortMsgLogger which will write into the logger
+func NewPortBandwidthLogger(
+	name string,
+	interval sim.VTimeInSec,
+	timeTeller sim.TimeTeller,
+	logger *log.Logger,
+) *PortBandwidthLogger {
+	h := new(PortBandwidthLogger)
+
+	h.name = name
+	h.logger = logger
+	h.TimeTeller = timeTeller
+	h.interval = interval
+	h.startTime = 0
+	h.endTime = h.CurrentTime()
+	h.messageCounter = make(map[sim.VTimeInSec]int64)
+	h.trafficBytesCounter = make(map[sim.VTimeInSec]int64)
+
+	return h
+}
+
+// Func writes the message information into the logger
+func (h *PortBandwidthLogger) Func(ctx sim.HookCtx) {
+	msg, ok := ctx.Item.(sim.Msg)
+	if !ok {
+		return
+	}
+
+	now := h.CurrentTime()
+
+	floorEndTime := sim.VTimeInSec(math.Round(math.Floor(float64(now/h.interval))*float64(h.interval)*1e10) / 1e10)
+	h.messageCount(floorEndTime)
+	h.trafficBytesCount(msg, floorEndTime)
+
+	diff := sim.VTimeInSec(math.Round(float64(floorEndTime-h.startTime)*1e10) / 1e10)
+
+	msgID := msg.Meta().ID
+	if msgID == "END" {
+		h.lastMsg(floorEndTime, ctx)
+	} else if diff >= 2*h.interval {
+		h.moreThanTwoInterval(floorEndTime, ctx)
+	} else if diff >= h.interval {
+		h.oneInterval(floorEndTime, ctx)
+	}
+}
+
+func (h *PortBandwidthLogger) messageCount(
```

Comment:
Use verbs for function names. So, I would rename it as `increaseMsgCount` and the next one `increaseTrafficByteCount`.